### ISSUE TITLE
[concept] requires only allows primary expressions

### ISSUE THREAD

**Issue by marehr** [ericniebler/range-v3#1138]
I'm currently trying to get a work-in-progress version of clang which supports concepts (see https://github.com/saarraz/clang-concepts) to work with the `v1.0-beta` version of ranges-v3.

After a bit of trial and error I encountered the following issue:

> https://en.cppreference.com/w/cpp/language/constraints#Requires_clauses
> The expression must have one of the following forms:
>    * a primary expression, e.g. Swappable<T>, std::is_integral<T>::value, (std::is_object_v<Args> && ...), or any parenthesized expression
>    * a sequence of primary expressions joined with the operator &&
>    * a sequence of aforementioned expressions joined with the operator ||

The ranges library is currently violating this on multiple occasions:

https://github.com/ericniebler/range-v3/blob/36c651faeb5c81bc28ac9ffcd97ad0602f2fade4/include/range/v3/iterator/common_iterator.hpp#L176-L182
`not ForwardIterator<I>` is not a primary expression, but `(not ForwardIterator<I>)` would be

https://github.com/ericniebler/range-v3/blob/36c651faeb5c81bc28ac9ffcd97ad0602f2fade4/include/range/v3/iterator/basic_iterator.hpp#L574-L582
same here

https://github.com/ericniebler/range-v3/blob/36c651faeb5c81bc28ac9ffcd97ad0602f2fade4/include/range/v3/detail/variant.hpp#L140-L145

`sizeof...(Ts) != 0` is not a primary expression

https://github.com/ericniebler/range-v3/blob/36c651faeb5c81bc28ac9ffcd97ad0602f2fade4/include/range/v3/detail/variant.hpp#L681-L686

`all_convertible_to<Args...>(0)` is not a primary expression

**Comment by ericniebler** [comment]
I'm sure there are many more such places, since I haven't tested with clang. Thanks for the report. Would you be willing to make a PR if you manage to get it working?

**Comment by marehr** [comment]
Yes, I can do, but I'm unsure how to proceed.

The easiest seems to wrap every `requires`-clause in parentheses, because this will make it a primary expression.
Another approach would be to keep the current expressions and wrap only non-primary expressions, such as`!literal`, in parenthesis.

**Comment by ericniebler** [comment]
> Another approach would be to keep the current expressions and wrap only non-primary expressions, such as `!literal`, in parenthesis.

This, please.

**Comment by marehr** [comment]
My PR (#1143) was merged that fixed a lot of these errors. But ranges is not completely building, because of this nasty bug (https://github.com/saarraz/clang-concepts/issues/53)

So should we close this issue or wait until that bug is fixed and reevaluate?

**Comment by CaseyCarter** [comment]
I worked around that particular occurrence in cmcstl2 by "hiding" the concept specialization prvalue behind a `bool` variable template. I suggest doing the same for range-v3:
```c++
namespace detail
{
#if defined(__clang__) && CPP_CXX_CONCEPTS // Workaround saarrraz/clang-concepts#53
    template<class R>
    RANGES_INLINE_VAR constexpr bool is_forwarding_range = ForwardingRange_<R>;

    template<class R, class U> requires Range<R>
    using maybe_dangling_ = if_then_t<__is_forwarding_range<R>, U, dangling>;
#else
    template<class R, class U> requires Range<R>
    using maybe_dangling_ = if_then_t<ForwardingRange_<R>, U, dangling>;
#endif
}
```

**Comment by marehr** [comment]
@CaseyCarter Should we open a new "meta" issue to track the progress of clang-concepts?

I applied your workaround and it works great. Should I create a PR to incorporate this?

I saw that there are `RANGES_WORKAROUND_*` macros in `/include/range/v3/detail/config.hpp`. Should I add a new one for this clang bug?

----

Furthermore I ran into a different kind of the same bug:
```c++
Scanning dependencies of target bug474
[  2%] Building CXX object test/CMakeFiles/bug474.dir/bug474.cpp.o
In file included from range-v3/test/bug474.cpp:11:
In file included from range-v3/include/range/v3/view/any_view.hpp:25:
range-v3/include/range/v3/view/facade.hpp:53:17: error: cannot yet mangle expression type ConceptSpecializationExpr
                Same<begin_cursor_t<Derived>, end_cursor_t<Derived>>,
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

https://github.com/ericniebler/range-v3/blob/45e4346fd2ed6e07e9822c28dd6837b4ca8de20b/include/range/v3/view/facade.hpp#L42-L47

When fixing it with

```c++
        template<typename Derived>
        RANGES_INLINE_VAR constexpr bool is_same_begin_end_cursor = Same<begin_cursor_t<Derived>, end_cursor_t<Derived>>;
        template<typename Derived>
        using facade_sentinel_t =
            meta::if_c<
                is_same_begin_end_cursor<Derived>,
                facade_iterator_t<Derived>,
                end_cursor_t<Derived>>;
```

I get 

```c++
[  2%] Building CXX object test/CMakeFiles/bug474.dir/bug474.cpp.o
In file included from range-v3/test/bug474.cpp:11:
In file included from range-v3/include/range/v3/view/any_view.hpp:25:
range-v3/include/range/v3/view/facade.hpp:33:38: error: no matching function for call to 'begin_cursor'
            detail::decay_t<decltype(range_access::begin_cursor(std::declval<Derived &>()))>;
                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
range-v3/include/range/v3/view/facade.hpp:44:74: note: in instantiation of template type alias 'begin_cursor_t'
      requested here
        RANGES_INLINE_VAR constexpr bool is_same_begin_end_cursor = Same<begin_cursor_t<Derived>, end_cursor_t<Derived>>;
                                                                         ^
range-v3/include/range/v3/view/facade.hpp:49:17: note: in instantiation of variable template specialization
      'ranges::detail::is_same_begin_end_cursor<const ranges::any_view<Foo &, ranges::category::input, void> >' requested here
                is_same_begin_end_cursor<Derived>,
                ^
range-v3/include/range/v3/view/facade.hpp:138:29: note: in instantiation of template type alias 'facade_sentinel_t'
      requested here
            CPP_ret(detail::facade_sentinel_t<D const>)(
                            ^
```

Maybe you have an Idea what happens.

**Comment by ericniebler** [comment]
I suggest (for native concepts):

```c++
        template<typename Derived>
        RANGES_INLINE_VAR constexpr bool is_same_begin_end_cursor = false;
        template<typename Derived>
            requires Same<begin_cursor_t<Derived>, end_cursor_t<Derived>>
        RANGES_INLINE_VAR constexpr bool is_same_begin_end_cursor<Derived> =true;

        template<typename Derived>
        using facade_sentinel_t =
            meta::if_c<
                is_same_begin_end_cursor<Derived>,
                facade_iterator_t<Derived>,
                end_cursor_t<Derived>>;
```

You are getting bitten by the fact that `Same<begin_cursor_t<Derived>, end_cursor_t<Derived>>` can fail either because the types are not the same, or because the types themselves are ill-formed, as it appears in your situation.

**Comment by marehr** [comment]
@ericniebler You are right! Your version works great.

The question remains:
>I applied your workaround and it works great. Should I create a PR to incorporate this?
>
> I saw that there are `RANGES_WORKAROUND_*` macros in `/include/range/v3/detail/config.hpp`. Should I add a new one for this clang bug?

**Comment by ericniebler** [comment]
Yes and yes! (Please, and thank you :-)

**Comment by marehr** [comment]
Thank you for fixing this!

@ericniebler Oh! I'm terribly sorry for not notifying you that I put this on backlog, because it turned out that even with all those workarounds that the experimental clang-concept branch did not compile in reasonable time and ram usage for us at that time. So we decided to make this not a priority and work on other stuff.

I think the workaround is not needed anymore, because saarraz/clang-concepts-old#53 said that it was fixed. We should still reevaluate it if it really fixed it.

I think we can close this issue and open a new one to track the progress on clang support.

Btw. did you try if ranges-v3 is compiling with visual studio? https://devblogs.microsoft.com/cppblog/c20-concepts-are-here-in-visual-studio-2019-version-16-3/

Considering that some things changed since when I would reevaluate the concept support of all major compilers.
